### PR TITLE
Implements memoization and introduces changes to the deployment scripts

### PR DESCRIPTION
**Pull request by lalithsuresh**
Implements memoization at the servers.

Nits to the deployment scripts.

### CODE REVIEW

File: deployment/demo-prepare.yaml [hunk 1]
```diff
@@ -6,10 +6,17 @@
...
+
     - name: Copy files
       copy: src={{ item.src }} dest={{ item.dest }}
       with_items:
               - {src: '../platform/target/data-cleaner-jar-with-dependencies.jar', dest: 'hillview'}
               - {src: '../data/On_Time.schema', dest: 'data'}
-    - name: Prepare datasets
-      shell: cd hillview; java -jar data-cleaner-jar-with-dependencies.jar
+              - {src: '../data/download.sh', dest: 'data'}
+              - {src: '.bashrc', dest: '.bashrc'}
```
Comment (mbudiu-vmw) [inline deployment/demo-prepare.yaml]: This is weird. Why do you need .bashrc and why do you expect one here?

File: platform/src/main/java/org/hillview/maps/LoadCsvFileMapper.java [hunk 2]
```diff
@@ -43,7 +43,7 @@ public LoadCsvFileMapper(final int which) {
 
         final List<CsvFileObject> result = new ArrayList<CsvFileObject>();
         if (this.which >= 0 && this.which <= 1) {
-            int limit = this.which == 0 ? 3 : 1;
+            int limit = this.which == 0 ? 12 : 1;
```
Comment (mbudiu-vmw) [inline platform/src/main/java/org/hillview/maps/LoadCsvFileMapper.java]: Maybe we should remove the limit altogether. I will do it later.

File: platform/src/main/java/org/hillview/remoting/HillviewServer.java [hunk 3]
```diff
@@ -17,21 +17,29 @@
...
  * Server that transfers map(), sketch(), zip() and unsubscribe() RPCs from a RemoteDataSet
  * object to locally managed IDataSet objects, and streams back results.
+ *
+ * If memoization is enabled, it caches the results of (operation, dataset-index) types.
  */
 public class HillviewServer extends HillviewServerGrpc.HillviewServerImplBase {
+    private static final Logger LOG = Logger.getLogger(HillviewServer.class.getName());
     public static final int DEFAULT_IDS_INDEX = 1;
     public static final int DEFAULT_PORT = 3569;
+    private static final boolean MEMOIZE = true;
```
Comment (mbudiu-vmw) [inline platform/src/main/java/org/hillview/remoting/HillviewServer.java]: Maybe this should not be final - so we can change it dynamically.

File: platform/src/main/java/org/hillview/remoting/HillviewServer.java [hunk 4]
```diff
@@ -79,8 +96,12 @@ public void onNext(final PartialResult<IDataSet> pr) {
                 }
                 final OperationResponse<Integer> res = new OperationResponse<Integer>(idsIndex);
                 final byte[] bytes = SerializationUtils.serialize(res);
-                responseObserver.onNext(PartialResponse.newBuilder()
-                                                       .setSerializedOp(ByteString.copyFrom(bytes)).build());
+                final PartialResponse result = PartialResponse.newBuilder()
+                        .setSerializedOp(ByteString.copyFrom(bytes)).build();
+                responseObserver.onNext(result);
+                if (MEMOIZE) {
+                    memoizedResult = result;
```
Comment (mbudiu-vmw) [inline platform/src/main/java/org/hillview/remoting/HillviewServer.java]: this.memoizedResult

File: platform/src/main/java/org/hillview/remoting/HillviewServer.java [hunk 5]
```diff
@@ -137,32 +171,50 @@ public void flatMap(final Command command, final StreamObserver<PartialResponse>
...
+                LOG.info("Returning memoized result for sketch operation against IDataSet#" + command.getIdsIndex());
+                return;
+            }
             final byte[] bytes = command.getSerializedOp().toByteArray();
             final SketchOperation sketchOp = SerializationUtils.deserialize(bytes);
             final Observable<PartialResult> observable = this.dataSets.get(command.getIdsIndex())
                                                                       .sketch(sketchOp.sketch);
+            final UUID commandId = new UUID(command.getHighId(), command.getLowId());
             final Subscription sub = observable.subscribe(new Subscriber<PartialResult>() {
+                @Nullable private Object result = sketchOp.sketch.getZero();
```
Comment (mbudiu-vmw) [inline platform/src/main/java/org/hillview/remoting/HillviewServer.java]: Let's call this one sketchResultAccumulator.
Also, use this. for referring to it.

File: platform/src/main/java/org/hillview/remoting/HillviewServer.java [hunk 6]
```diff
@@ -240,4 +297,18 @@ private boolean checkValidIdsIndex(final int index,
...
+    private boolean respondIfReplyIsMemoized(final Command command,
+                                             final StreamObserver<PartialResponse> responseObserver) {
+        if (MEMOIZE && memoizedCommands.containsKey(command.getSerializedOp())
+             && memoizedCommands.get(command.getSerializedOp()).containsKey(command.getIdsIndex())) {
+            responseObserver.onNext(memoizedCommands.get(command.getSerializedOp()).get(command.getIdsIndex()));
+            responseObserver.onCompleted();
+            return true;
+        }
+        return false;
+    }
```
Comment (mbudiu-vmw) [inline platform/src/main/java/org/hillview/remoting/HillviewServer.java]: Can you also please add a method to purge the cache?
No callers for it yet.

**Comment by lalithsuresh** [comment]
Addresses your comments.

**Pull request merged**